src/esp/gfx/Renderer.cpp

```diff
@@ -74,6 +70,51 @@ struct Renderer::Impl {
   void draw(RenderCamera& camera, MagnumDrawableGroup& drawables) {
     renderEnter();
     camera.getMagnumCamera().setViewport(framebufferSize_);
+
+    /* Inverted projection matrix to unproject the depth value and chop the
+       near plane off. We don't care about X/Y there and the corresponding
+       parts of the matrix are zero as well so take just the lower-right part
+       of it (denoted a, b, c, d).
+
+        x 0 0 0
+        0 y 0 0
+        0 0 a b
+        0 0 c d
+
+      Taking a 2-component vector with the first component being Z and second
+      1, the final calculation of unprojected Z is then
+
+        | a b |   | z |   | az + b |
+        | c d | * | 1 | = | cz + d |
+
+    */
+    const Matrix4 projection = camera.getMagnumCamera().projectionMatrix();
```

Comment:
You could potentially save some floating point operations by converting to Matrix2x2 here.

Reply:
that's exactly what I'm doing here?

Reply:
Yes. I was thinking you could create a projection2x2 matrix from projection and deal with 2x2 from then. So drop into 2x2 operations even sooner.

```
const Matrix2x2 projection2x2 = Matrix2x2{Math::swizzle<'z', 'w'>(projection[2]), Math::swizzle<'z', 'w'>(projection[3])};
```

Reply:
Ah, I see. I was also experimenting with doing this whole inverse calculation only if the projection changes (which is basically never) instead of every frame, but I didn't see any measurable improvement either -- I guess it's a statistical error compared to all the ops done on a quarter of a megapixel :)

In the end I reverted the caching and strived to keep this precalculation code as straightforward as possible. Proving an inverse of a 2x2 gives the same result as 4x4 wouldn't be straightforward, for me at least.

Reply:
This is a block diagonal matrix, so the inverse rules are know: https://en.wikipedia.org/wiki/Block_matrix#Block_matrix_inversion  :)

Reply:
Ahhh thank you :) That freshened up my memory, now it's obvious.